File: Tests/AINotebookCoreTests/IngestionServiceTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class IngestionServiceTests: XCTestCase {

    func testIngestPlainTextEndToEnd() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")

        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("ing-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: dir) }
        let file = dir.appendingPathComponent("memo.txt")
        try "Hello world. Second sentence.".write(to: file, atomically: true, encoding: .utf8)

        let service = IngestionService(store: store)
        let source = try await service.ingestFile(file, into: nb.id!)

        // Refresh status from disk
        let reloaded = try XCTUnwrap(try store.source(id: source.id!))
        XCTAssertEqual(reloaded.status, .ready)
        XCTAssertEqual(reloaded.type, .text)
        XCTAssertEqual(reloaded.title, "memo")

        let chunks = try store.chunks(sourceId: source.id!)
        XCTAssertGreaterThan(chunks.count, 0)
        XCTAssertEqual(chunks.first?.ord, 0)
    }

    func testIngestRawTextCreatesPersistedSource() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")

        let service = IngestionService(store: store)
        let source = try await service.ingestRawText(
            title: "My note",
            text: String(repeating: "lorem ipsum ", count: 500),
            into: nb.id!
        )
        XCTAssertEqual(source.status, .ready)
        XCTAssertEqual(source.type, .text)
        let chunks = try store.chunks(sourceId: source.id!)
        XCTAssertGreaterThan(chunks.count, 1)
    }

    func testIngestUnknownExtensionLeavesSourceInErrorStatus() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")

        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("ing-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: dir) }
        let file = dir.appendingPathComponent("mystery.bin")
        try Data([0x01, 0x02, 0x03]).write(to: file)

        let service = IngestionService(store: store)
        do {
            _ = try await service.ingestFile(file, into: nb.id!)
            XCTFail("expected throw for unsupported extension")
        } catch IngestionService.IngestionError.unsupportedExtension {
            // ok — no source row should have been created
            XCTAssertEqual(try store.sources(notebookId: nb.id!).count, 0)
        } catch {
            XCTFail("wrong error: \(error)")
        }
    }
}

File: windows/tests/AINotebook.Core.Tests/Ingestion/IngestionServiceTests.cs
using AINotebook.Core;
using AINotebook.Core.Ingestion;
using AINotebook.Core.Models;
using AINotebook.Core.Storage;
using Xunit;

namespace AINotebook.Core.Tests.Ingestion;

public class IngestionServiceTests
{
    [Fact]
    public async Task IngestPlainTextEndToEnd()
    {
        using var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("NB");

        var dir = Path.Combine(Path.GetTempPath(), "ing-" + Guid.NewGuid());
        Directory.CreateDirectory(dir);
        try
        {
            var file = Path.Combine(dir, "memo.txt");
            File.WriteAllText(file, "Hello world. Second sentence.");

            var service = new IngestionService(store);
            var source = await service.IngestFileAsync(new Uri(file), nb.Id!.Value);

            // Refresh status from disk.
            var reloaded = store.Source(source.Id!.Value);
            Assert.NotNull(reloaded);
            Assert.Equal(SourceStatus.Ready, reloaded!.Status);
            Assert.Equal(SourceType.Text, reloaded.Type);
            Assert.Equal("memo", reloaded.Title);

            var chunks = store.Chunks(source.Id!.Value);
            Assert.True(chunks.Count > 0);
            Assert.Equal(0, chunks[0].Ord);
        }
        finally
        {
            Directory.Delete(dir, recursive: true);
        }
    }

    [Fact]
    public async Task IngestRawTextCreatesPersistedSource()
    {
        using var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("NB");

        var service = new IngestionService(store);
        var source = await service.IngestRawTextAsync(
            "My note",
            string.Concat(Enumerable.Repeat("lorem ipsum ", 500)),
            nb.Id!.Value);

        Assert.Equal(SourceStatus.Ready, source.Status);
        Assert.Equal(SourceType.Text, source.Type);
        var chunks = store.Chunks(source.Id!.Value);
        Assert.True(chunks.Count > 1);
    }

    [Fact]
    public async Task IngestUnknownExtensionLeavesNoSourceRow()
    {
        using var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("NB");

        var dir = Path.Combine(Path.GetTempPath(), "ing-" + Guid.NewGuid());
        Directory.CreateDirectory(dir);
        try
        {
            var file = Path.Combine(dir, "mystery.bin");
            File.WriteAllBytes(file, new byte[] { 0x01, 0x02, 0x03 });

            var service = new IngestionService(store);
            await Assert.ThrowsAsync<IngestionException.UnsupportedExtension>(
                () => service.IngestFileAsync(new Uri(file), nb.Id!.Value));

            // No source row should have been created.
            Assert.Empty(store.Sources(nb.Id!.Value));
        }
        finally
        {
            Directory.Delete(dir, recursive: true);
        }
    }
}
